src/main/java/com/o19s/es/ltr/feature/store/PrecompiledExpressionFeature.java

```diff
@@ -65,13 +76,41 @@ public String name() {
 
     @Override
     public Query doToQuery(LtrQueryContext context, FeatureSet set, Map<String, Object> params) {
-        return new DerivedExpressionQuery(set, expression);
+        List<String> missingParams = queryParams.stream()
+                .filter((x) -> params == null || !params.containsKey(x))
+                .collect(Collectors.toList());
+        if (!missingParams.isEmpty()) {
+            String names = missingParams.stream().collect(Collectors.joining(","));
+            throw new IllegalArgumentException("Missing required param(s): [" + names + "]");
+        }
+
+        Map<String, Double> queryParamValues = getQueryParamValues(params);
+        return new DerivedExpressionQuery(set, expression, queryParamValues);
+    }
+
+    /* if a param is an expression variable then we need to use it as a `variable` in lucene expression exposed via Bindings */
+    private Map<String, Double> getQueryParamValues(Map<String, Object> params) {
+        Map<String, Double> queryParamValues = new HashMap<>();
+        for (String param : queryParams) {
+            if (expressionVariables.contains(param)) {
+                try {
+                    queryParamValues.put(param, (Double) params.get(param));
```

Comment:
casting like that seems a bit strict to me. I'd do some testing but at a glance I'd cast to Number then call `Number::doubleValue()`.
Reason is that you may receive a int for queries like:
`"param": 1`
And possibly accept String using Double.parseDouble() if you want to be lenient (I don't have strong opinions on this).

But overall I'd adjust and assert all possible formats using unit tests.

Reply:
makes sense. I now cast to `Number`